### ISSUE TITLE
Quickstart example with different NN libraries does not tag correctly

### ISSUE THREAD

**Issue by Binbose** [google-deepmind/kfac-jax#114]
Hey,

I am adapting the quickstart example to equinox to use it in my project. However, it seems only the bias is correctly tagged (with Auto[scale_and_shift_tag_0]); the weight matrix is tagged as 'Orphan'. I also tried it with flax, in which case only the weight matrix is tagged correctly (with Auto[dense_tag_0]), but the bias is tagged as 'Orphan'.  And lastly, for pure jax it seems again only to tag the bias correctly. 

I am not sure what I am doing wrong here; I introduced minimal changes. In the test script below the four different libraries can be switched between using the `lib_type` variable in the code. I would appreciate any help.

```
import haiku as hk
import jax
import jax.numpy as jnp
import kfac_jax
from absl import logging
import sys
import equinox as eqx
import flax
import jax.random as random

logging.get_absl_handler().python_handler.stream = sys.stdout
logging.set_verbosity(logging.INFO)

# Hyper parameters
NUM_CLASSES = 10
L2_REG = 1e-3
NUM_BATCHES = 5
NUM_FEATURES = 20
rng = jax.random.PRNGKey(42)


def make_dataset_iterator(batch_size):
  # Dummy dataset, in practice this should be your dataset pipeline
  for _ in range(NUM_BATCHES):
    yield jnp.zeros([batch_size, NUM_FEATURES]), jnp.ones([batch_size], dtype="int32")


def softmax_cross_entropy(logits: jnp.ndarray, targets: jnp.ndarray):
  """Softmax cross entropy loss."""
  # We assume integer labels
  assert logits.ndim == targets.ndim + 1

  # Tell KFAC-JAX this model represents a classifier
  # See https://kfac-jax.readthedocs.io/en/latest/overview.html#supported-losses
  kfac_jax.register_softmax_cross_entropy_loss(logits, targets)
  log_p = jax.nn.log_softmax(logits, axis=-1)
  return - jax.vmap(lambda x, y: x[y])(log_p, targets)


lib_type = 'hk' # 'hk', 'eqx', 'flax' 'jax'
#######################################################################################################################
#######################################################################################################################
#######################################################################################################################
if lib_type == 'hk':
  def model_fn(x):
    return hk.nets.MLP(
      output_sizes=(50, 50, NUM_CLASSES),
      with_bias=True,
      activation=jax.nn.tanh,
    )(x)

  hk_model = hk.without_apply_rng(hk.transform(model_fn))

  def loss_fn(model_params, model_batch):
    """The loss function to optimize."""
    x, y = model_batch
    logits = hk_model.apply(model_params, x)
    loss = jnp.mean(softmax_cross_entropy(logits, y))

    return loss + L2_REG * kfac_jax.utils.inner_product(params, params) / 2.0

  optimizer = kfac_jax.Optimizer(
    value_and_grad_func=jax.value_and_grad(loss_fn),
    l2_reg=L2_REG,
    value_func_has_aux=False,
    value_func_has_state=False,
    value_func_has_rng=False,
    use_adaptive_learning_rate=True,
    use_adaptive_momentum=True,
    use_adaptive_damping=True,
    initial_damping=1.0,
    multi_device=False,
  )

  input_dataset = make_dataset_iterator(128)
  dummy_images, dummy_labels = next(input_dataset)
  rng, key = jax.random.split(rng)
  params = hk_model.init(key, dummy_images)
  rng, key = jax.random.split(rng)
  opt_state = optimizer.init(params, key, (dummy_images, dummy_labels))

#######################################################################################################################
#######################################################################################################################
#######################################################################################################################
elif lib_type == 'eqx':
  class simple_net(eqx.Module):
      net: callable

      def __init__(self, key=None):
        keys = jax.random.split(key,  2)
        self.net = eqx.nn.MLP(NUM_FEATURES, NUM_CLASSES, 100, 0, activation=jax.nn.tanh, key=keys[1])

      def __call__(self, x):
        return self.net(x)

  eqx_model = simple_net(rng)
  params, static = eqx.partition(eqx_model, eqx.is_inexact_array)


  def loss_fn(model_params, static, model_batch):
    """The loss function to optimize."""
    x, y = model_batch
    eqx_model = eqx.combine(model_params, static)
    logits = jax.vmap(eqx_model)(x)
    loss = jnp.mean(softmax_cross_entropy(logits, y))

    return loss + L2_REG * kfac_jax.utils.inner_product(params, params) / 2.0
  kfac_loss_fn = lambda params, batch: loss_fn(params, static, batch)

  optimizer = kfac_jax.Optimizer(
    value_and_grad_func=jax.value_and_grad(kfac_loss_fn),
    l2_reg=L2_REG,
    value_func_has_aux=False,
    value_func_has_state=False,
    value_func_has_rng=False,
    use_adaptive_learning_rate=True,
    use_adaptive_momentum=True,
    use_adaptive_damping=True,
    initial_damping=1.0,
    multi_device=False,
  )

  input_dataset = make_dataset_iterator(128)
  dummy_images, dummy_labels = next(input_dataset)
  rng, key = jax.random.split(rng)
  opt_state = optimizer.init(params, key, (dummy_images, dummy_labels))

#######################################################################################################################
#######################################################################################################################
#######################################################################################################################
elif lib_type == 'flax':
  class MLP(flax.linen.Module):

    def setup(self):
      self.dense1 = flax.linen.Dense(32)
      self.dense2 = flax.linen.Dense(NUM_CLASSES)

    def __call__(self, x):
      x = self.dense1(x)
      x = flax.linen.relu(x)
      x = self.dense2(x)
      return x


  flax_model = MLP()
  params = flax_model.init(rng, jnp.zeros([128, NUM_FEATURES]))

  def loss_fn(model_params, model_batch):
    """The loss function to optimize."""
    x, y = model_batch
    logits = flax_model.apply(model_params, x)
    loss = jnp.mean(softmax_cross_entropy(logits, y))

    return loss + L2_REG * kfac_jax.utils.inner_product(params, params) / 2.0

  optimizer = kfac_jax.Optimizer(
    value_and_grad_func=jax.value_and_grad(loss_fn),
    l2_reg=L2_REG,
    value_func_has_aux=False,
    value_func_has_state=False,
    value_func_has_rng=False,
    use_adaptive_learning_rate=True,
    use_adaptive_momentum=True,
    use_adaptive_damping=True,
    initial_damping=1.0,
    multi_device=False,
  )

  input_dataset = make_dataset_iterator(128)
  dummy_images, dummy_labels = next(input_dataset)
  rng, key = jax.random.split(rng)
  opt_state = optimizer.init(params, key, (dummy_images, dummy_labels))

#######################################################################################################################
#######################################################################################################################
#######################################################################################################################
elif lib_type == 'jax':
  def random_layer_params(m, n, key, scale=1e-2):
    w_key, b_key = random.split(key)
    return scale * random.normal(w_key, (n, m)), scale * random.normal(b_key, (n,))


  # Initialize all layers for a fully-connected neural network with sizes "sizes"
  def init_network_params(sizes, key):
    keys = random.split(key, len(sizes))
    return [random_layer_params(m, n, k) for m, n, k in zip(sizes[:-1], sizes[1:], keys)]

  def relu(x):
    return jnp.maximum(0, x)

  def predict(params, x):
    activations = x
    for w, b in params[:-1]:
      outputs = jnp.dot(w, activations) + b
      activations = relu(outputs)

    final_w, final_b = params[-1]
    logits = jnp.dot(final_w, activations) + final_b
    return logits


  params = init_network_params([NUM_FEATURES, 100, NUM_CLASSES], random.PRNGKey(0))


  def loss_fn(model_params, model_batch):
    """The loss function to optimize."""
    x, y = model_batch
    logits = jax.vmap(predict, in_axes=(None, 0))(model_params, x)
    loss = jnp.mean(softmax_cross_entropy(logits, y))

    return loss + L2_REG * kfac_jax.utils.inner_product(params, params) / 2.0

  optimizer = kfac_jax.Optimizer(
    value_and_grad_func=jax.value_and_grad(loss_fn),
    l2_reg=L2_REG,
    value_func_has_aux=False,
    value_func_has_state=False,
    value_func_has_rng=False,
    use_adaptive_learning_rate=True,
    use_adaptive_momentum=True,
    use_adaptive_damping=True,
    initial_damping=1.0,
    multi_device=False,
  )

  input_dataset = make_dataset_iterator(128)
  dummy_images, dummy_labels = next(input_dataset)
  rng, key = jax.random.split(rng)
  opt_state = optimizer.init(params, key, (dummy_images, dummy_labels))


# Training loop
for i, batch in enumerate(input_dataset):
  rng, key = jax.random.split(rng)
  params, opt_state, stats = optimizer.step(
      params, opt_state, key, batch=batch, global_step_int=i)
  print(i, stats)
```

**Comment by botev** [comment]
I will take a look when I have the time. Technically, for such issues you might want to register whatever is the computation for your layers, similar to how the haiku layers are registered.

**Comment by Binbose** [comment]
Jey, thank you for the answer! The jaxpr are (only the NN, without the loss function part):

For equinox
```
{ lambda a:f32[10,20] b:f32[10]; c:f32[128,20]. let
    d:f32[10,128] = dot_general[dimension_numbers=(([1], [1]), ([], []))] a c
    e:f32[128,10] = transpose[permutation=(1, 0)] d
    f:f32[1,10] = broadcast_in_dim[broadcast_dimensions=(1,) shape=(1, 10)] b
    g:f32[128,10] = add e f
  in (g,) }
```

For flax:
```
{ lambda ; a:f32[32] b:f32[20,32] c:f32[10] d:f32[32,10] e:f32[128,20]. let
    f:f32[128,32] = dot_general[dimension_numbers=(([1], [0]), ([], []))] e b
    g:f32[1,32] = reshape[dimensions=None new_sizes=(1, 32)] a
    h:f32[128,32] = add f g
    i:f32[128,32] = custom_jvp_call[
      call_jaxpr={ lambda ; j:f32[128,32]. let
          k:f32[128,32] = pjit[
            jaxpr={ lambda ; l:f32[128,32]. let
                m:f32[128,32] = max l 0.0
              in (m,) }
            name=relu
          ] j
        in (k,) }
      jvp_jaxpr_thunk=<function _memoize.<locals>.memoized at 0x7f96f078beb0>
      num_consts=0
      symbolic_zeros=False
    ] h
    n:f32[128,10] = dot_general[dimension_numbers=(([1], [0]), ([], []))] i d
    o:f32[1,10] = reshape[dimensions=None new_sizes=(1, 10)] c
    p:f32[128,10] = add n o
  in (p,) }
```

For pure jax
```
{ lambda ; a:f32[100,20] b:f32[100] c:f32[10,100] d:f32[10] e:f32[128,20]. let
    f:f32[100,128] = dot_general[dimension_numbers=(([1], [1]), ([], []))] a e
    g:f32[128,100] = transpose[permutation=(1, 0)] f
    h:f32[1,100] = broadcast_in_dim[broadcast_dimensions=(1,) shape=(1, 100)] b
    i:f32[128,100] = add g h
    j:f32[128,100] = max 0.0 i
    k:f32[10,128] = dot_general[dimension_numbers=(([1], [1]), ([], []))] c j
    l:f32[128,10] = transpose[permutation=(1, 0)] k
    m:f32[1,10] = broadcast_in_dim[broadcast_dimensions=(1,) shape=(1, 10)] d
    n:f32[128,10] = add l m
  in (n,) }
```

**Comment by botev** [comment]
The issue for equinox and Jax is that you are taking a transpose in the middle, which would probably mess up some of the curvature blocks as well. E.g. from what I can see you have input `(128, 20)`, w `(100, 20)` and bias `(100,)` and you are computing `x W^T + b`. This would require additional things to make it compatible since KFAC assumes `x W + b` and this makes a difference since now you've change which axis of the weight matrix corresponds to inputs and which one to outputs.

For `flax` I will add support for that and show you how to do it yourself, the reason why it doesn't work automatically is that they are reshaping the bias, rather than broadcasting.

**Comment by botev** [comment]
Can you try and see if the commit fixes your flax issue?

**Comment by Binbose** [comment]
I just installed the newest version from github but flax unfortunately still doesn't work (btw, I think there is a typo in the commit, `_dense_with_reshae` instead of `_dense_with_reshape`)
The pure Jax version works now, thank you!
For equinox, does it mean we either have to create custom equinox MLP layers or implement a new dense_tag for the standard ones?

**Comment by botev** [comment]
Thanks for catching the typo.

For Equinox its difficult because of the transposition. I do think that if you implement a custom MLP where the matmul is done in the correct way, e.g. `x @ W + b` it should work.

**Comment by botev** [comment]
I'm closing this unless there is more to this?
